cmd/optimizer/main.go

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright (c) 2023. Nydus Developers. All rights reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package main
+
+import (
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"sigs.k8s.io/yaml"
+
+	"github.com/containerd/nri/pkg/api"
+	"github.com/containerd/nri/pkg/stub"
+	"github.com/containerd/nydus-snapshotter/pkg/fanotify"
+)
+
+type config struct {
+	LogFile string   `yaml:"logFile"`
+	Events  []string `yaml:"events"`
+
+	ServerPath string `yaml:"serverPath"`
+	PersistDir string `yaml:"persistDir"`
+	Timeout    int    `yaml:"timeout"`
+	Overwrite  bool   `yaml:"overwrite"`
+}
+
+type plugin struct {
+	stub stub.Stub
+	mask stub.EventMask
+}
+
+var (
+	cfg                  config
+	log                  *logrus.Logger
+	_                    = stub.ConfigureInterface(&plugin{})
+	globalFanotifyServer = make(map[string]*fanotify.Server)
+)
+
+const (
+	imageNameLabel = "io.kubernetes.cri.image-name"
+)
+
+func (p *plugin) Configure(config, runtime, version string) (stub.EventMask, error) {
+	log.Infof("got configuration data: %q from runtime %s %s", config, runtime, version)
+	if config == "" {
+		return p.mask, nil
+	}
+
+	oldCfg := cfg
+	err := yaml.Unmarshal([]byte(config), &cfg)
+	if err != nil {
+		return 0, fmt.Errorf("failed to parse provided configuration: %w", err)
+	}
+
+	p.mask, err = api.ParseEventMask(cfg.Events...)
+	if err != nil {
+		return 0, fmt.Errorf("failed to parse events in configuration: %w", err)
+	}
+
+	if cfg.LogFile != oldCfg.LogFile {
```

Comment:
When to close the file `f`

Reply:
Now the optimizer server will close the file descriptor after sending the `path` to the client.

Reply:
Emm... It is not the same issue. The file descriptor mentioned in last comment is created by optimizer server. So we close it after getting the path.

The file `f` is closed before exiting this NRI plugin, i.e. in `onClose` event.